### ISSUE TITLE
[Coloring] Highlighting unused declarations / open statements stops working

### ISSUE THREAD

**Issue by vasily-kirichenko** [fsprojects/VisualFSharpPowerTools#1014]
At some point Unused stops working - the margin is not updated (it has outdated markers), new unused symbols are not highlighted as Unused.

Debugger showed that background compilation is triggered OK https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs#L329 but the callback https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs#L339 is not, hence the issue. 

I'm struggling to understand why FSC stopped calling the callback? Maybe the callback throws an exception and FSC classified it as "malicious" and removed from its list?

This never happened before your yesterday optimization.

**Comment by dungpa** [comment]
The callback at https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs#L339 is only called when the project is compiled.

Is the `updateUnusedDeclarations` function called? We hook it at `ProjectChecked` event at https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs#L357.

I'm not sure why it stops working at all.

**Comment by vasily-kirichenko** [comment]
Yes, I was wrong, I mean this callback https://github.com/fsprojects/VisualFSharpPowerTools/blob/master/src/FSharpVSPowerTools.Logic/SyntaxConstructClassifier.fs#L357

It's not called. I try to add `protect` for the whole body of it and will wait for exceptions.

**Comment by vasily-kirichenko** [comment]
It stopped working again. The only errors in our log are (4 duplicates):
```
[VFPT][Error 01:18:15 ] Context: Exception occurs while querying lexing states.
Exception Message: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
Stack Trace:    at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at Microsoft.VisualStudio.Editor.Implementation.LanguageServiceClassificationTagger.TryGetStartingStateForLine(Int32 lineNo)
   at Microsoft.VisualStudio.Editor.Implementation.TextDocData.GetColorStateAtStartOfLine(Int32 iLine, Int32& piState)
   at Microsoft.VisualStudio.FSharp.LanguageService.VsTextColorState.GetColorStateAtStartOfLine(IVsTextColorState colorState, Int32 line)
   at lambda_method(Closure , IVsTextColorState , Int32 )
   at FSharpVSPowerTools.ProjectSystem.VSLanguageService.buildQueryLexState(ITextBuffer textBuffer, String source, FSharpList`1 defines, Int32 line)
```

**Comment by dungpa** [comment]
I'm trying to fix it at #1015. I think it has the same root cause as #1009 does i.e. things have been invoked on a wrong thread.

**Comment by vasily-kirichenko** [comment]
Fixed in https://github.com/fsprojects/VisualFSharpPowerTools/pull/1015
